# Capstone Project - The Battle of the Neighborhoods (Week 1 and 2)
### Applied Data Science Capstone by IBM/Coursera

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)



## Introduction: Business Problem <a name="introduction"></a>

Healing-Touch is a famous pharmaceutical company that wish to establish a 'Medical Supply Store' somewhere in Bronx Area. In this project we will try to find an optimal location for a 'Medical Supply'. Specifically, this report will be targeted to Healing-Touch Pharmaceuticals.

Bronx may have some 'Medical Supply stores' in some neighbourhoods already. So our target would be to get an overview about the locations having lack of Medical Supply stores.

We will use  data science to generate a few most  neighborhoods based on this criteria and suggestions can be made to client based on our analysis

## Data <a name="data"></a>

Based on definition of our problem, factors that will influence our decission are:
* presence of existing Medical Supplies in the neighborhood


Following data sources will be needed to extract/generate the required information:
* Coordinate of Bronx, NY neighburhood can be extracted from the available newyork json data
* Coordinate of Bronx, NY city centre can be generated by using Google Maps API geocoding
* All kind of Veneues in every neighborhood will be obtained using **Foursquare API**


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.8-main

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2021.5.30          |   py38h578d9bd_0         141 KB  conda-forge
    geographiclib-1.52         |     pyhd8ed1ab_0          35 KB  conda-forge
    geopy-2.2.0                |     pyhd8ed1ab_0          67 KB  conda-forge
    openssl-1.1.1k             |       h7f98852_0         2.1 MB  conda-forge
    python_abi-3.8             |           2_cp38           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.4 MB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.52-pyhd8ed1ab_0
  geopy              conda-forge/noarch::geopy-2.2.0-pyhd8ed1ab_0
  pyth

## 1. Download and Explore Dataset


In [4]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


In [5]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [6]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [9]:
neighborhoods_data = newyork_data['features']

In [17]:

neighborhoods_data[0]


{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the data into a _pandas_ dataframe

In [10]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Then let's loop through the data and fill the dataframe one row at a time.

In [11]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [38]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [39]:
Bronx_data = neighborhoods[neighborhoods['Borough'] == 'Bronx'].reset_index(drop=True)
Bronx_data.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

#### Let's first find the latitude & longitude of Bronx city center, using specific, well known address and Google Maps geocoding API.

In [40]:
address = 'Bronx, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Bronx are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Bronx are 40.8466508, -73.8785937.


Let's visualizat Bronx with the neighborhoods in it.

In [41]:
# create map of Bronx using latitude and longitude values
map_Bronx = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Bronx_data['Latitude'], Bronx_data['Longitude'], Bronx_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=10,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Bronx)  
    
map_Bronx

### Foursquare
Now that we have our location candidates, let's use Foursquare API to get info on Venuews in each neighborhood.

We're interested in venues named 'Medical Supplies'

#### Define Foursquare Credentials and Version

In [14]:
CLIENT_ID = 'MU1KAFNUNAZUM4WHF5LJVFEOGJUXIF1YCC30PVQGMCOIIIOG' # your Foursquare ID
CLIENT_SECRET = 'ZSPILD0X5KX3DOUKKD514DBLAXZYB203GZXYVFVGLMOCPX2N' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MU1KAFNUNAZUM4WHF5LJVFEOGJUXIF1YCC30PVQGMCOIIIOG
CLIENT_SECRET:ZSPILD0X5KX3DOUKKD514DBLAXZYB203GZXYVFVGLMOCPX2N


## 2. Explore Neighborhoods in Bronx


In [65]:
#### Let's create a function to repeat the same process to all the neighborhoods in Bronx

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called _Bronx_venues_.

In [42]:
Bronx_venues = getNearbyVenues(names=Bronx_data['Neighborhood'],
                                   latitudes=Bronx_data['Latitude'],
                                   longitudes=Bronx_data['Longitude']
                                  )

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Claremont Village
Concourse Village
Mount Eden
Mount Hope
Bronxdale
Allerton
Kingsbridge Heights


#### Let's check the size of the resulting dataframe

In [43]:
print(Bronx_venues.shape)
Bronx_venues.head()

(1234, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

              Venue  Venue Latitude  Venue Longitude  Venue Category  
0  Lollipops Gelato       40.894123       -73.845892    Dessert Shop  
1          Rite Aid       40.896649       -73.844846        Pharmacy  
2         Walgreens       40.896528       -73.844700        Pharmacy  
3  Carvel Ice Cream       40.890487       -73.848568  Ice Cream Shop  
4           Dunkin'       40.890459       -73.849089      Donut Shop

Let's check how many venues were returned for each neighborhood

In [44]:
Bronx_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Allerton                                32                      32     32   
Baychester                              24                      24     24   
Bedford Park                            31                      31     31   
Belmont                                100                     100    100   
Bronxdale                               15                      15     15   
Castle Hill                              7                       7      7   
City Island                             29                      29     29   
Claremont Village                       22                      22     22   
Clason Point                            13                      13     13   
Co-op City                              17                      17     17   
Concourse                               24                      24     24   
Concourse Village                       42                      42     42   
Country Club                             6                       6      6   
East Tremont                            20                      20     20   
Eastchester                             23                      23     23   
Edenwald                                 5                       5      5   
Edgewater Park                          20                      20     20   
Fieldston                                4                       4      4   
Fordham                                 82                      82     82   
High  Bridge                            24                      24     24   
Hunts Point                             14                      14     14   
Kingsbridge                             70                      70     70   
Kingsbridge Heights                     37                      37     37   
Longwood                                 9                       9      9   
Melrose                                 30                      30     30   
Morris Heights                           7                       7      7   
Morris Park                             21                      21     21   
Morrisania                              27                      27     27   
Mott Haven                              19                      19     19   
Mount Eden                              37                      37     37   
Mount Hope                              16                      16     16   
North Riverdale                         24                      24     24   
Norwood                                 32                      32     32   
Olinville                                9                       9      9   
Parkchester                             29                      29     29   
Pelham Bay                              42                      42     42   
Pelham Gardens                          18                      18     18   
Pelham Parkway                          23                      23     23   
Port Morris                             12                      12     12   
Riverdale                               12                      12     12   
Schuylerville                           17                      17     17   
Soundview                               16                      16     16   
Spuyten Duyvil                          12                      12     12   
Throgs Neck                             11                      11     11   
Unionport                               23                      23     23   
University Heights                      20                      20     20   
Van Nest                                13                      13     13   
Wakefield                                9                       9      9   
West Farms                              22                      22     22   
Westchester Square                      34                      34     34   
Williamsbridge   

#### Let's find out how many unique categories can be curated from all the returned venues


In [45]:
print('There are {} uniques categories.'.format(len(Bronx_venues['Venue Category'].unique())))

There are 167 uniques categories.


## 3. Analyze Each Neighborhood

In [46]:
# one hot encoding
Bronx_onehot = pd.get_dummies(Bronx_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Bronx_onehot['Neighborhood'] = Bronx_venues['Neighborhood'] 



In [42]:
# move neighborhood column to the first column
fixed_columns = [Bronx_onehot.columns[-1]] + list(Bronx_onehot.columns[:-1])
Bronx_onehot = Bronx_onehot[fixed_columns]

Bronx_onehot.head()

Neighborhood  Accessories Store  Acupuncturist  Adult Boutique  \
0  Marble Hill                  0              0               0   
1  Marble Hill                  0              0               0   
2  Marble Hill                  0              0               0   
3  Marble Hill                  0              0               0   
4  Marble Hill                  0              0               0   

   Afghan Restaurant  African Restaurant  American Restaurant  Antique Shop  \
0                  0                   0                    0             0   
1                  0                   0                    0             0   
2                  0                   0                    0             0   
3                  0                   0                    0             0   
4                  0                   0                    0             0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  Auto Workshop  BBQ Joint  \
0                      0                    0              0          0   
1                      0                    0              0          0   
2                      0                    0              0          0   
3                      0                    0              0          0   
4                      0                    0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0           0           0       0     0    0               0   
1           0           0       0     0    0               0   
2           0           0       0     0    0               0   
3           0           0       0     0    0               0   
4           0           0       0     0    0               0   

   Basketball Court  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                 0         0            0           0              0   
1                 0         0            0           0              0   
2                 0         0            0           0              0   
3                 0         0            0           0              0   
4                 0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [47]:
Bronx_grouped = Bronx_onehot.groupby('Neighborhood').mean().reset_index()
Bronx_grouped

Neighborhood       ATM  Accessories Store  African Restaurant  \
0              Allerton  0.000000           0.000000            0.000000   
1            Baychester  0.000000           0.000000            0.000000   
2          Bedford Park  0.000000           0.000000            0.000000   
3               Belmont  0.000000           0.000000            0.000000   
4             Bronxdale  0.000000           0.000000            0.000000   
5           Castle Hill  0.000000           0.000000            0.000000   
6           City Island  0.000000           0.000000            0.000000   
7     Claremont Village  0.000000           0.000000            0.000000   
8          Clason Point  0.000000           0.000000            0.000000   
9            Co-op City  0.000000           0.000000            0.000000   
10            Concourse  0.000000           0.000000            0.000000   
11    Concourse Village  0.000000           0.000000            0.000000   
12         Country Club  0.000000           0.000000            0.000000   
13         East Tremont  0.000000           0.000000            0.000000   
14          Eastchester  0.000000           0.000000            0.000000   
15             Edenwald  0.000000           0.000000            0.000000   
16       Edgewater Park  0.000000           0.000000            0.000000   
17            Fieldston  0.000000           0.000000            0.000000   
18              Fordham  0.012195           0.012195            0.012195   
19         High  Bridge  0.000000           0.000000            0.000000   
20          Hunts Point  0.000000           0.000000            0.000000   
21          Kingsbridge  0.000000           0.000000            0.000000   
22  Kingsbridge Heights  0.000000           0.000000            0.000000   
23             Longwood  0.000000           0.000000            0.000000   
24              Melrose  0.000000           0.000000            0.000000   
25       Morris Heights  0.000000           0.000000            0.000000   
26          Morris Park  0.000000           0.000000            0.000000   
27           Morrisania  0.000000           0.000000            0.000000   
28           Mott Haven  0.000000           0.000000            0.000000   
29           Mount Eden  0.000000           0.000000            0.000000   
30           Mount Hope  0.000000           0.000000            0.000000   
31      North Riverdale  0.000000           0.000000            0.000000   
32              Norwood  0.000000           0.000000            0.000000   
33            Olinville  0.000000           0.000000            0.000000   
34          Parkchester  0.000000           0.000000            0.000000   
35           Pelham Bay  0.000000           0.000000            0.000000   
36       Pelham Gardens  0.000000           0.000000            0.000000   
37       Pelham Parkway  0.000000           0.000000            0.000000   
38          Port Morris  0.000000           0.000000            0.000000   
39            Riverdale  0.000000           0.000000            0.000000   
40        Schuylerville  0.000000           0.000000            0.000000   
41            Soundview  0.000000           0.000000            0.000000   
42       Spuyten Duyvil  0.000000           0.000000            0.000000   
43          Throgs Neck  0.000000           0.000000            0.000000   
44            Unionport  0.000000           0.000000            0.000000   
45   University Heights  0.000000           0.000000            0.000000   
46             Van Nest  0.000000           0.000000            0.000000   
47            Wakefield  0.000000           0.000000            0.000000   
48           West Farms  0.000000           0.000000            0.000000   
49   Westchester Square  0.000000           0.000000            0.000000   
50       Williamsbridge  0.000000           0.000000            0.000000   
51             Woodlawn  0.000000           0.000000       

#### Let's put that into a _pandas_ dataframe

First, let's write a function to sort the venues in ascending order.

In [67]:
def return_least_common_venues(row, num_bottom_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=True)
    
    return row_categories_sorted.index.values[0:num_bottom_venues]

Now let's create the new dataframe and display the least occuring 10 venues for each neighborhood.

In [68]:
num_bottom_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of bottom venues
columns = ['Neighborhood']
for ind in np.arange(num_bottom_venues):
    try:
        columns.append('{}{} Least Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Least Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Bronx_grouped['Neighborhood']

for ind in np.arange(Bronx_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_least_common_venues(Bronx_grouped.iloc[ind, :], num_bottom_venues)

neighborhoods_venues_sorted

Neighborhood 1st Least Common Venue    2nd Least Common Venue  \
0              Allerton                    ATM      Medical Supply Store   
1            Baychester                    ATM      Medical Supply Store   
2          Bedford Park                    ATM                    Market   
3               Belmont                    ATM            Mattress Store   
4             Bronxdale                    ATM      Medical Supply Store   
5           Castle Hill                    ATM               Men's Store   
6           City Island                    ATM  Mediterranean Restaurant   
7     Claremont Village                    ATM               Men's Store   
8          Clason Point                    ATM            Mattress Store   
9            Co-op City                    ATM      Medical Supply Store   
10            Concourse                    ATM  Mediterranean Restaurant   
11    Concourse Village                    ATM            Mattress Store   
12         Country Club                    ATM               Men's Store   
13         East Tremont                    ATM            Mattress Store   
14          Eastchester                    ATM               Men's Store   
15             Edenwald                    ATM               Men's Store   
16       Edgewater Park                    ATM  Mediterranean Restaurant   
17            Fieldston                    ATM             Metro Station   
18              Fordham         Ice Cream Shop       Martial Arts School   
19         High  Bridge                    ATM      Medical Supply Store   
20          Hunts Point                    ATM      Medical Supply Store   
21          Kingsbridge                    ATM                Laundromat   
22  Kingsbridge Heights                    ATM       Martial Arts School   
23             Longwood                    ATM               Men's Store   
24              Melrose                    ATM            Mattress Store   
25       Morris Heights                    ATM               Men's Store   
26          Morris Park                    ATM  Mediterranean Restaurant   
27           Morrisania                    ATM               Men's Store   
28           Mott Haven                    ATM      Medical Supply Store   
29           Mount Eden                    ATM       Martial Arts School   
30           Mount Hope                    ATM       Martial Arts School   
31      North Riverdale                    ATM      Medical Supply Store   
32              Norwood                    ATM            Mattress Store   
33            Olinville                    ATM             Metro Station   
34          Parkchester                    ATM                    Lounge   
35           Pelham Bay                    ATM      Medical Supply Store   
36       Pelham Gardens                    ATM  Mediterranean Restaurant   
37       Pelham Parkway                    ATM      Medical Supply Store   
38          Port Morris                    ATM      Medical Supply Store   
39            Riverdale                    ATM               Men's Store   
40        Schuylerville                    ATM  Mediterranean Restaurant   
41            Soundview                    ATM               Men's Store   
42       Spuyten Duyvil                    ATM            Mattress Store   
43          Throgs Neck                    ATM             Metro Station   
44            Unionport                    ATM               Men's Store   
45   University Heights                    ATM               Men's Store   
46             Van Nest                    ATM               Men's Store   
47            Wakefield                    ATM             Metro Station   
48           West Farms                    ATM  Mediterranean Restaurant   
49   Westchester Square                    ATM                    Lounge   
50       Williamsbridge                    ATM               Men's Store   
51             Woodlawn                    ATM            M

## 4. Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.

In [69]:
# set number of clusters
kclusters = 5

Bronx_grouped_clustering = Bronx_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Bronx_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:] 

array([1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 4, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 0, 1, 3, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the bottom 10 venues for each neighborhood.

In [70]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Bronx_merged = Bronx_data

# merge Bronx_grouped with Bronx_data to add latitude/longitude for each neighborhood
Bronx_merged = Bronx_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Bronx_merged.head()

Borough Neighborhood   Latitude  Longitude  Cluster Labels  \
0   Bronx    Wakefield  40.894705 -73.847201               0   
1   Bronx   Co-op City  40.874294 -73.829939               0   
2   Bronx  Eastchester  40.887556 -73.827806               0   
3   Bronx    Fieldston  40.895437 -73.905643               4   
4   Bronx    Riverdale  40.890834 -73.912585               0   

  1st Least Common Venue 2nd Least Common Venue    3rd Least Common Venue  \
0                    ATM          Metro Station        Mexican Restaurant   
1                    ATM   Medical Supply Store  Mediterranean Restaurant   
2                    ATM            Men's Store             Metro Station   
3                    ATM          Metro Station        Mexican Restaurant   
4                    ATM            Men's Store             Metro Station   

  4th Least Common Venue 5th Least Common Venue 6th Least Common Venue  \
0     Miscellaneous Shop      Mobile Phone Shop          Moving Target   
1            Men's Store          Metro Station     Mexican Restaurant   
2     Mexican Restaurant     Miscellaneous Shop      Mobile Phone Shop   
3     Miscellaneous Shop      Mobile Phone Shop          Moving Target   
4     Mexican Restaurant     Miscellaneous Shop      Mobile Phone Shop   

  7th Least Common Venue 8th Least Common Venue    9th Least Common Venue  \
0            Music Venue              Nightclub               Men's Store   
1     Miscellaneous Shop      Mobile Phone Shop            Mattress Store   
2          Moving Target            Music Venue  Mediterranean Restaurant   
3            Music Venue              Nightclub               Men's Store   
4          Moving Target            Music Venue  Mediterranean Restaurant   

  10th Least Common Venue  
0   Outdoors & Recreation  
1           Moving Target  
2               Nightclub  
3   Outdoors & Recreation  
4               Nightclub

Finally, let's visualize the resulting clusters

In [72]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Bronx_merged['Latitude'], Bronx_merged['Longitude'], Bronx_merged['Neighborhood'], Bronx_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Methodology <a name="methodology"></a>

In this project we will direct our efforts on detecting areas of Bronx that have low Medical Supplies density.

In first step we have collected the least occuring venues for all the neighbourhoods in Bronx, NY

Second step in our analysis will be to find out the neighbourhoods having least frequency of Medical Supplies

In third and final step we will focus on most promising areas for exploration and search for optimal venue location by stakeholders.

## Analysis <a name="analysis"></a>

Let's perform some basic explanatory data analysis and derive some additional info from our raw data. First let's find out the neighbourhoods having Medical Supplies as the least occuring venue in bottom 2

In [87]:
Bronx_Medical=Bronx_merged.iloc[:,0:7].drop(['Cluster Labels'],axis=1)
Bronx_Medical

Borough         Neighborhood   Latitude  Longitude 1st Least Common Venue  \
0    Bronx            Wakefield  40.894705 -73.847201                    ATM   
1    Bronx           Co-op City  40.874294 -73.829939                    ATM   
2    Bronx          Eastchester  40.887556 -73.827806                    ATM   
3    Bronx            Fieldston  40.895437 -73.905643                    ATM   
4    Bronx            Riverdale  40.890834 -73.912585                    ATM   
5    Bronx          Kingsbridge  40.881687 -73.902818                    ATM   
6    Bronx             Woodlawn  40.898273 -73.867315                    ATM   
7    Bronx              Norwood  40.877224 -73.879391                    ATM   
8    Bronx       Williamsbridge  40.881039 -73.857446                    ATM   
9    Bronx           Baychester  40.866858 -73.835798                    ATM   
10   Bronx       Pelham Parkway  40.857413 -73.854756                    ATM   
11   Bronx          City Island  40.847247 -73.786488                    ATM   
12   Bronx         Bedford Park  40.870185 -73.885512                    ATM   
13   Bronx   University Heights  40.855727 -73.910416                    ATM   
14   Bronx       Morris Heights  40.847898 -73.919672                    ATM   
15   Bronx              Fordham  40.860997 -73.896427         Ice Cream Shop   
16   Bronx         East Tremont  40.842696 -73.887356                    ATM   
17   Bronx           West Farms  40.839475 -73.877745                    ATM   
18   Bronx         High  Bridge  40.836623 -73.926102                    ATM   
19   Bronx              Melrose  40.819754 -73.909422                    ATM   
20   Bronx           Mott Haven  40.806239 -73.916100                    ATM   
21   Bronx          Port Morris  40.801664 -73.913221                    ATM   
22   Bronx             Longwood  40.815099 -73.895788                    ATM   
23   Bronx          Hunts Point  40.809730 -73.883315                    ATM   
24   Bronx           Morrisania  40.823592 -73.901506                    ATM   
25   Bronx            Soundview  40.821012 -73.865746                    ATM   
26   Bronx         Clason Point  40.806551 -73.854144                    ATM   
27   Bronx          Throgs Neck  40.815109 -73.816350                    ATM   
28   Bronx         Country Club  40.844246 -73.824099                    ATM   
29   Bronx          Parkchester  40.837938 -73.856003                    ATM   
30   Bronx   Westchester Square  40.840619 -73.842194                    ATM   
31   Bronx             Van Nest  40.843608 -73.866299                    ATM   
32   Bronx          Morris Park  40.847549 -73.850402                    ATM   
33   Bronx              Belmont  40.857277 -73.888452                    ATM   
34   Bronx       Spuyten Duyvil  40.881395 -73.917190                    ATM   
35   Bronx      North Riverdale  40.908543 -73.904531                    ATM   
36   Bronx           Pelham Bay  40.850641 -73.832074                    ATM   
37   Bronx        Schuylerville  40.826580 -73.826203                    ATM   
38   Bronx       Edgewater Park  40.821986 -73.813885                    ATM   
39   Bronx          Castle Hill  40.819014 -73.848027                    ATM   
40   Bronx            Olinville  40.871371 -73.863324                    ATM   
41   Bronx       Pelham Gardens  40.862966 -73.841612                    ATM   
42   Bronx            Concourse  40.834284 -73.915589                    ATM   
43   Bronx            Unionport  40.829774 -73.850535                    ATM   
44   Bronx             Edenwald  40.884561 -73.848083                    ATM   
45   Bronx    Claremont Village  40.831428 -73.901199                    ATM   
46   Bronx    Concourse Village  40.824780 -73.915847                    ATM   
47   Bronx           Mount Eden  40.843826 -73.916556                    ATM   
48   Bronx           Mount Hope  40.848842 -73.908299                    ATM   
49 

We shall include only the neighbourhoods  having 'Medical Supply Store' in any above

In [171]:
Bronx_Medical1 = Bronx_Medical[Bronx_Medical['2nd Least Common Venue']=='Medical Supply Store']

In [174]:
Bronx_Medical1

Borough     Neighborhood   Latitude  Longitude 1st Least Common Venue  \
1    Bronx       Co-op City  40.874294 -73.829939                    ATM   
9    Bronx       Baychester  40.866858 -73.835798                    ATM   
10   Bronx   Pelham Parkway  40.857413 -73.854756                    ATM   
18   Bronx     High  Bridge  40.836623 -73.926102                    ATM   
20   Bronx       Mott Haven  40.806239 -73.916100                    ATM   
21   Bronx      Port Morris  40.801664 -73.913221                    ATM   
23   Bronx      Hunts Point  40.809730 -73.883315                    ATM   
35   Bronx  North Riverdale  40.908543 -73.904531                    ATM   
36   Bronx       Pelham Bay  40.850641 -73.832074                    ATM   
49   Bronx        Bronxdale  40.852723 -73.861726                    ATM   
50   Bronx         Allerton  40.865788 -73.859319                    ATM   

   2nd Least Common Venue  
1    Medical Supply Store  
9    Medical Supply Store  
10   Medical Supply Store  
18   Medical Supply Store  
20   Medical Supply Store  
21   Medical Supply Store  
23   Medical Supply Store  
35   Medical Supply Store  
36   Medical Supply Store  
49   Medical Supply Store  
50   Medical Supply Store

In [91]:
!pip install haversine

In [172]:
import haversine as hs # for calculating the distance betweeen 2 points with Coordinates available



Calculate the distance of each neighbourhood from Bronx City Centre

In [169]:
print('--------------------Distance from City Centre-------------------------\n')
print('------Neighbourhood------------------------Distance from Bronx Centre (kms)------\n')

for t in range(len(Bronx_Medical1)):
    latsrt=Bronx_Medical1.iloc[t, 2]
    lonsrt=Bronx_Medical1.iloc[t, 3]
    loc2= (latsrt,lonsrt)
    distan=round(hs.haversine(loc2,Bronx_loc),2)
    print(' ' * 7 , Bronx_Medical1.iloc[t, 1], '  ' * 20 ,  distan)
    


--------------------Distance from City Centre-------------------------

------Neighbourhood------------------------Distance from Bronx Centre (kms)------

        Co-op City                                          5.12
        Baychester                                          4.24
        Pelham Parkway                                          2.33
        High  Bridge                                          4.15
        Mott Haven                                          5.49
        Port Morris                                          5.79
        Hunts Point                                          4.12
        North Riverdale                                          7.22
        Pelham Bay                                          3.94
        Bronxdale                                          1.57
        Allerton                                          2.68


## Results and Discussion <a name="results"></a>

Our analysis shows that although there are quite many neighbourhoods in Bronx where lack of Medical Supply can be found, we have shortlisted the list where 'Medical Supply' is the least found venue.

After directing our attention to this more narrow area of interest , we found out the distance between the shortlisted neighbourhoods and the Bronx city centre.

Result of all this is 11 neighbourhoods with the distance from city centre ranging between 1.5 to 6 kms. Client can now be provided with these options for creating a 'Medical Supply Centre'

## Conclusion <a name="conclusion"></a>

Purpose of this project was to identify Bronx neighbourhood close to center with low number of 'Medical Supply Centre'in order to aid stakeholders in narrowing down the search for optimal location. By calculating 'Medical Supply density' density distribution from Foursquare data we have first identified neighbourhood that justify further analysis by stakeholders.

Final decission on optimal 'Medical Supply Centre' location will be made by stakeholders based on specific characteristics of neighborhoods and locations in every recommended zone, taking into consideration additional factors like social and economic dynamics of every neighborhood, convenience of transportation, etc.